File: src/data/courses/dsa/trees/binary-tree-core-operations/max-min-value/code/python/max_min_value.py
# ============================================================
# Max and Min Value in Binary Tree
# ============================================================

class Solution1:
    def findMax(self, root) -> int:
        if not root: return float('-inf')
        return max(root.val, self.findMax(root.left), self.findMax(root.right))

    def findMin(self, root) -> int:
        if not root: return float('inf')
        return min(root.val, self.findMin(root.left), self.findMin(root.right))

class Solution2:
    def findBSTMin(self, root) -> int:
        while root.left: root = root.left
        return root.val

    def findBSTMax(self, root) -> int:
        while root.right: root = root.right
        return root.val

File: src/data/courses/dsa/trees/binary-tree-core-operations/max-min-value/code/java/MaxMinValue.java
// ============================================================
// Max and Min Value in Binary Tree
// ============================================================

class Solution1 {
    public int findMax(TreeNode root) {
        if (root == null) return Integer.MIN_VALUE;
        return Math.max(root.val, Math.max(findMax(root.left), findMax(root.right)));
    }
    public int findMin(TreeNode root) {
        if (root == null) return Integer.MAX_VALUE;
        return Math.min(root.val, Math.min(findMin(root.left), findMin(root.right)));
    }
}

class Solution2 {
    public int findBSTMin(TreeNode root) { while (root.left!=null) root=root.left; return root.val; }
    public int findBSTMax(TreeNode root) { while (root.right!=null) root=root.right; return root.val; }
}
